pkg/skaffold/build/deps.go

```diff
@@ -0,0 +1,114 @@
+package build
+
+import (
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/skaffold/pkg/skaffold/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type DependencyMap struct {
+	artifacts       []*config.Artifact
+	pathToArtifacts map[string][]*config.Artifact
+}
+
+type DependencyResolver interface {
+	GetDependencies(a *config.Artifact) ([]string, error)
+}
+
+//TODO(@r2d4): Figure out best UX to support configuring this blacklist
+var ignoredPrefixes = []string{"vendor", ".git"}
+
+func (d *DependencyMap) Paths() []string {
+	allPaths := []string{}
+	for path := range d.pathToArtifacts {
+		allPaths = append(allPaths, path)
+	}
+	sort.Strings(allPaths)
+	return allPaths
+}
+
+func (d *DependencyMap) ArtifactsForPaths(paths []string) []*config.Artifact {
+	m := map[*config.Artifact]struct{}{}
+	for _, p := range paths {
+		artifacts := d.pathToArtifacts[p]
+		for _, a := range artifacts {
+			m[a] = struct{}{}
+		}
+	}
+	artifacts := []*config.Artifact{}
+	for a := range m {
+		artifacts = append(artifacts, a)
+	}
+	return artifacts
+}
+
+func NewDependencyMap(artifacts []*config.Artifact, res DependencyResolver) (*DependencyMap, error) {
+	m, err := pathToArtifactMap(artifacts, res)
+	if err != nil {
+		return nil, errors.Wrap(err, "generating path to artifact map")
+	}
+	return &DependencyMap{
+		artifacts:       artifacts,
+		pathToArtifacts: m,
+	}, nil
+}
+
+// path must be an absolute path
+func isIgnored(workspace, path string) (bool, error) {
+	for _, ig := range ignoredPrefixes {
+		ignoredPrefix, err := filepath.Abs(filepath.Join(workspace, ig))
+		if err != nil {
+			return false, errors.Wrapf(err, "calculating absolute path of ignored dep %s", ig)
+		}
+
+		if strings.HasPrefix(path, ignoredPrefix) {
+			logrus.Debugf("Ignoring %s for artifact dependencies", path)
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func pathToArtifactMap(artifacts []*config.Artifact, res DependencyResolver) (map[string][]*config.Artifact, error) {
+	m := map[string][]*config.Artifact{}
+	for _, a := range artifacts {
+		paths, err := pathsForArtifact(a, res)
+		if err != nil {
+			return nil, errors.Wrapf(err, "getting paths for artifact %s", a.DockerfilePath)
+		}
+		for _, p := range paths {
+			artifacts, ok := m[p]
+			if !ok {
+				m[p] = []*config.Artifact{a}
+				continue
+			}
+			m[p] = append(artifacts, a)
```

Comment:
Can this be either artifacts = append(artifacts, a) or m[p] = append(m[p], a) for consistency?